### DISCUSSION TITLE
Tabs, x⃣ button with many tabs open, tab resizing, opened tabs list, tab groups

### DISCUSSION THREAD

**Discussion by WWWFWWW** [mikekazakov/nimble-commander#133] [Ideas]
Just a few ideas about tabs and their management:

1) Tabs resizing and close tab button

With many, many tabs open in a panel, the tabs' names get quite understandably truncated to just ... or less, which makes the hover‑over‑tab  x⃣ "close tab" button somewhat "dangerous", as it's easier to misclick and close the tab instead of selecting it.

![TempScreenshot](https://github.com/mikekazakov/nimble-commander/assets/30300299/ecf47b6b-90cd-42e8-bb74-d9ddcd002d5e)

It's the same issue Safari and other browsers have when having many tabs open (I know, I shouldn't, but...), where user's misclicking easily closes a tab instead of selecting it. Made a bit worse as if you open an insane amount of tabs, they get resized next to nothing and the close tab button takes up almost all the space in the tab (though that's an extreme case).

I wonder if an option to disable "close tab button" on the _inactive_ tabs when truncated below a certain size would be doable? If not adding much overhead, as the tabs are apparently resized dynamically.

2) Dynamic resizing of tabs might be better with some lowest fixed size limit, which brings me to and would probably require:

3) Tab list

Would be really nice to have an _opened tabs_ list, either as a Quick List or even a special panel (like Processes). A dropdown button showing all opened tabs in each panel when clicked would be even more intuitive. You already have a +⃣ button for New Tab, so maybe a  ㆔⃣ button on the opposite side of it or something like that?

4) Tab groups

Favourites are great, but even more friendly in the long run would be tab or panel _groups_, similar to tab groups in latest Safari.

One usually works on projects with several tabs open in each panel, relevant to just that specific project. Having a saveable tab group that remembers which tabs were open in each panel and could be saved, restored or seamlessly switched to a different group when one switches projects.

That last one is obviously a major undertaking, as it would require a lot of thought on the implementation (e.g. should a saved tab group be static, what opened windows should be restored upon restart, warning on user closing a window with a group when changed, et cetera). Though not ideal, something like Safari 17 attempts with their tab groups.

**Comment by mikekazakov** [comment]
> Just a few ideas about tabs and their management:
> 
> 1. Tabs resizing and close tab button
> 
> With many, many tabs open in a panel, the tabs' names get quite understandably truncated to just ... or less, which makes the hover‑over‑tab x⃣ "close tab" button somewhat "dangerous", as it's easier to misclick and close the tab instead of selecting it.
> 
> ![TempScreenshot](https://private-user-images.githubusercontent.com/30300299/301173833-ecf47b6b-90cd-42e8-bb74-d9ddcd002d5e.JPG?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDY5NzUyNzcsIm5iZiI6MTcwNjk3NDk3NywicGF0aCI6Ii8zMDMwMDI5OS8zMDExNzM4MzMtZWNmNDdiNmItOTBjZC00MmU4LWJiNzQtZDlkZGNkMDAyZDVlLkpQRz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMDMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjAzVDE1NDI1N1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTg3MjU3ZjExYWVmN2E4YjliYzJlODg3MWJhMzczOTUyZTk3MWZiZTgzNThkMDk5MzJlMDJmY2MzYTJhNmEzZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.Wv-T5XhGVq_Xpw1cLDR7n-UygV-lgVr88oj9C7r_M-A)
> 
> It's the same issue Safari and other browsers have when having many tabs open (I know, I shouldn't, but...), where user's misclicking easily closes a tab instead of selecting it. Made a bit worse as if you open an insane amount of tabs, they get resized next to nothing and the close tab button takes up almost all the space in the tab (though that's an extreme case).
> 
> I wonder if an option to disable "close tab button" on the _inactive_ tabs when truncated below a certain size would be doable? If not adding much overhead, as the tabs are apparently resized dynamically.

Good idea, yes, I think this should be doable. 

> 2. Dynamic resizing of tabs might be better with some lowest fixed size limit, which brings me to and would probably require:

Image seem to be missing?

> 3. Tab list
> 
> Would be really nice to have an _opened tabs_ list, either as a Quick List or even a special panel (like Processes). A dropdown button showing all opened tabs in each panel when clicked would be even more intuitive. You already have a +⃣ button for New Tab, so maybe a ㆔⃣ button on the opposite side of it or something like that?

There's something similar in NC already: F1/F2 shows the locations of all tabs at the bottom-most section on the pop-up menu.

> 4. Tab groups
> 
> Favourites are great, but even more friendly in the long run would be tab or panel _groups_, similar to tab groups in latest Safari.
> 
> One usually works on projects with several tabs open in each panel, relevant to just that specific project. Having a saveable tab group that remembers which tabs were open in each panel and could be saved, restored or seamlessly switched to a different group when one switches projects.
> 
> That last one is obviously a major undertaking, as it would require a lot of thought on the implementation (e.g. should a saved tab group be static, what opened windows should be restored upon restart, warning on user closing a window with a group when changed, et cetera). Though not ideal, something like Safari 17 attempts with their tab groups.

Yes, fully agree, a combination of favourites with groups would be very useful.

To be clear - the library NC currently uses for tab management is rather old (~18 years old) and is a bit outdated nowadays. I'd be happy to switch to something more modern, either third-party or written from scratch within NC, but that's a quite significant undertaking.

**Comment by WWWFWWW** [comment]
Re 2.: Nope, no image missing there, the colon was just a followup to the latter point 3. I just meant that there is no minimum tab size, so opening an insane amount of tabs (which I sometimes do, in the absence of saveable tab groups, though not that insanely big as in the screenshot below) might in an extreme case look something like this:
![TempScreenshot](https://github.com/mikekazakov/nimble-commander/assets/30300299/518bd06a-6316-475b-9598-4ba8daf7bad2)
(note that in the right panel, the tabs don't even fit a single character of text, while in the left one, the name of the tabs gets truncated such that it shows just part of the character).

Re 3.: 

> There's something similar in NC already: F1/F2 shows the locations of all tabs at the bottom-most section on the pop-up menu.

I seem to be missing that on 1.5, is there some option for that I might have missed? My F1/F2 are mapped to Go/Left Panel... and right respectively, which I believe is the default shortcut mapping? It looks like this:
![aaa](https://github.com/mikekazakov/nimble-commander/assets/30300299/fd4c47c9-a33f-4ef6-9c0a-556d4b831a5a)
The only locations showing below the favourites are mounted disks, in this case just the system one.

Re 4.: Bummer, so even the above UI issues might be just hard‑baked into the library used? That might not even allow doing something about my point 2 if it doesn't allow fixed minimum tab sizes or such, and just resizes them to fit all in the window.

Though at least the old lib seems to be pretty robust – I just tried to make it crash by opening an insane amount of tabs of the same folder by holding ⌘‑T for several minutes, opening probably like several thousand tabs or so, and it held very well even as the the tabs rendered were just a single pixel wide, still not running out of memory of anything. I did make it crash _eventually_ after a while, with a few more thousands of them, but that's something that wouldn't really happen in any even remotely imaginable real use scenario (I was just fuzzing it and curious :) ).

But yes, I totally get that switching to a different tab UI library would be quite an undertaking. And I am afraid I am not really proficient enough to try coding it myself, sorry ;-)

Cheers, my best regards and thanks – NC is still my go‑to orthodox file commander for Mac, IMHO the best one out there.
